pinot-core/src/main/java/org/apache/pinot/core/operator/query/SelectionPartiallyOrderedByOperator.java

```
@@ -0,0 +1,303 @@
+   * by that comparator.
+   *
+   * There is a second comparator called {@link #_unsortedComparator} that is used to sort rows inside each partition.
+   * When calling {@link #add(Object[])} with a row that doesn't belong to the current partition, a new one is started
+   * and the previous one is sorted. Therefore, this object maintains the invariant that at any moment there are at
+   * least {@link #sortedSize()} elements that are completely sorted, which means that no matter which elements are
+   * added after that, these elements are going to be smallest.
+   */
+  private static class PartiallySortedListBuilder {
+    private final ArrayList<Object[]> _rows;
+    private final Comparator<Object[]> _sortedComparator;
+    private final Comparator<Object[]> _unsortedComparator;
+
+    public PartiallySortedListBuilder(int expectedSize, Comparator<Object[]> sortedComparator,
+        Comparator<Object[]> unsortedComparator) {
+      _rows = new ArrayList<>(expectedSize);
+      _sortedComparator = sortedComparator;
+      _unsortedComparator = unsortedComparator;
+    }
+
+    @Nullable
+    private Object[] _lastBlockRow;
+    private int _lastSorted;
+
+    /**
+     * Adds the given row to this object. The new column must be equal or higher than the last added row in relation to
+     * {@link #_sortedComparator}.
+     *
+     * @param row The row to add. The values of the already sorted columns must be equal or higher than the last added
+     *           row, if any.
+     * @return true if and only if the previous block was closed.
+     */
+    boolean add(Object[] row) {
+      boolean result;
+      if (_lastBlockRow == null) {
+        _lastSorted = 0;
+        _lastBlockRow = row;
+        result = false;
+      } else {
+        int cmp = _sortedComparator.compare(_lastBlockRow, row);
+        assert cmp >= 0 : "Row " + Arrays.toString(row) + " is lower than previously added row"
```

Consider using Preconditions class?

---

This is more a "computable comment" than something I would like to evaluate every time one element is being added.

---

This assertion was changed with an explicit `if` given that in case of failure, the `AssetionError` is not caught by Pinot job, so the error is lost.